File: app/src/main/java/vn/ditagis/com/tanhoa/qlsc/services/GetDMA.java
package vn.ditagis.com.tanhoa.qlsc.services;

import android.content.Context;
import android.os.AsyncTask;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.Feature;
import com.esri.arcgisruntime.data.FeatureQueryResult;
import com.esri.arcgisruntime.data.QueryParameters;
import com.esri.arcgisruntime.data.ServiceFeatureTable;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ExecutionException;

import vn.ditagis.com.tanhoa.qlsc.R;
import vn.ditagis.com.tanhoa.qlsc.entities.DLayerInfo;
import vn.ditagis.com.tanhoa.qlsc.entities.entitiesDB.ListObjectDB;


public class GetDMA extends AsyncTask<Void, Boolean, Void> {
    private Context mContext;

    private AsyncResponse mDelegate;

    public interface AsyncResponse {

        void processFinish();
    }

    public GetDMA(Context context, AsyncResponse delegate) {
        this.mContext = context;
        this.mDelegate = delegate;
    }


    private void getMaDMAFromService() {

        String layerInfoVatTu = mContext.getString(R.string.LayerInfo_DMA);

        for (DLayerInfo dLayerInfo : ListObjectDB.getInstance().getLstFeatureLayerDTG()) {
            if (dLayerInfo.getId().equals(layerInfoVatTu)) {
                final QueryParameters queryParameters = new QueryParameters();
                queryParameters.setWhereClause("1=1");
                String url = dLayerInfo.getUrl();
                if (!url.startsWith("http"))
                    url = "http:" + dLayerInfo.getUrl();
                final ServiceFeatureTable serviceFeatureTable = new ServiceFeatureTable(url);
//
                final ListenableFuture<FeatureQueryResult> feature = serviceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL);
                feature.addDoneListener(() -> {
                    final List<String> dmaList = new ArrayList<>();
                    try {
                        FeatureQueryResult result = feature.get();
                        Iterator<Feature> iterator = result.iterator();
                        Feature item;
                        while (iterator.hasNext()) {
                            item = iterator.next();
                            Object dma = item.getAttributes().get(mContext.getString(R.string.field_DMA_maDMA));
                            if (dma != null)
                                dmaList.add(dma.toString());
                        }
                        ListObjectDB.getInstance().setDmas(dmaList);
                        publishProgress(true);

                    } catch (InterruptedException | ExecutionException e) {
                        e.printStackTrace();
                        publishProgress();
                    }


                });

                break;
            }
        }

    }


    @Override
    protected Void doInBackground(Void... voids) {
        getMaDMAFromService();
        return null;
    }

    @Override
    protected void onProgressUpdate(Boolean... values) {
        super.onProgressUpdate(values);
        if(values!= null && values.length >0 && values[0])
            mDelegate.processFinish();
    }
}

File: app/src/main/java/vn/ditagis/com/tanhoa/qlsc/services/GetDMA.kt
package vn.ditagis.com.tanhoa.qlsc.services

import android.os.AsyncTask

import com.esri.arcgisruntime.data.Feature
import com.esri.arcgisruntime.data.QueryParameters
import com.esri.arcgisruntime.data.ServiceFeatureTable

import java.util.ArrayList
import java.util.concurrent.ExecutionException

import vn.ditagis.com.tanhoa.qlsc.entities.Constant
import vn.ditagis.com.tanhoa.qlsc.entities.DApplication
import vn.ditagis.com.tanhoa.qlsc.entities.entitiesDB.ListObjectDB


class GetDMA(private val mApplication: DApplication, private val mDelegate: AsyncResponse) : AsyncTask<Void, Boolean, Void>() {

    interface AsyncResponse {

        fun processFinish()
    }


    private fun getMaDMAFromService() {

        val layerInfoVatTu = Constant.IDLayer.DMA
        for (dLayerInfo in mApplication.lstFeatureLayerDTG!!) {
            if (dLayerInfo.id == layerInfoVatTu) {
                val queryParameters = QueryParameters()
                queryParameters.whereClause = "1=1"
                var url = dLayerInfo.url
                if (!url.startsWith("http"))
                    url = "http:" + dLayerInfo.url
                val serviceFeatureTable = ServiceFeatureTable(url)
                //
                val feature = serviceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL)
                feature.addDoneListener {
                    val dmaList = ArrayList<String>()
                    try {
                        val result = feature.get()
                        val iterator = result.iterator()
                        var item: Feature
                        while (iterator.hasNext()) {
                            item = iterator.next()
                            val dma = item.attributes[Constant.FieldDMA.MA_DMA]
                            if (dma != null)
                                dmaList.add(dma.toString())
                        }
                        ListObjectDB.instance.dmas = dmaList
                        publishProgress(true)

                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                        publishProgress()
                    } catch (e: ExecutionException) {
                        e.printStackTrace()
                        publishProgress()
                    }


                }

                break
            }
        }

    }


    override fun doInBackground(vararg voids: Void): Void? {
        getMaDMAFromService()
        return null
    }

    override fun onProgressUpdate(vararg values: Boolean?) {
        super.onProgressUpdate(*values)
        if (values.isNotEmpty() && values[0]!!)
            mDelegate.processFinish()
    }
}
